File: 프로그래머스/3/12927. 야근 지수/야근 지수.cpp
#include <bits/stdc++.h>

using namespace std;

priority_queue<int> pq;

long long solution(int n, vector<int> works) {
    long long answer = 0;
    
    for(auto cur : works){
        pq.push(cur);
    }
    
    while(n-- > 0){
        auto cur = pq.top(); pq.pop();
        if(cur == 0) break;
        pq.push(cur-1);
    }
    
    while(!pq.empty()){
        auto cur = pq.top(); pq.pop();
        answer += cur * cur;
    }
    
    return answer;
}

File: 프로그래머스/3/12927. 야근 지수/야근 지수.java
import java.util.*;

class Solution {
    public long solution(int n, int[] works) {
        
        //작업량 제곱의 합 = 야근 피로도
        //1시간 = 작업량 1
        //퇴근까지 남은 시간 -> 각 일에 대한 작업량의 피로도 최솟값 -> 골고루 해야 하겠네
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        Integer a[] = Arrays.stream(works).boxed().toArray(Integer[]::new); 
        
        for (Integer work : a) {
            pq.add(work);
        }
        
        for (int i = 0; i < n; i++){
            Integer temp = pq.poll();
            if (temp == 0) {
                break;
            }
            temp--;
            pq.add(temp);
        }
        
        long result = 0;
        
        while (pq.size() != 0) {
            Integer temp = pq.poll();
            result += temp * temp;
        }
        
        return result;
    }
}